server/src/main/java/org/elasticsearch/injection/Injector.java

```diff
@@ -0,0 +1,360 @@
+     *     </li>
+     *     <li>
+     *         A constructor parameter of some other class being instantiated,
+     *         having exactly the right class (not a supertype)
+     *     </li>
+     * </ol>
+     *
+     * @return <code>this</code>
+     */
+    public Injector addClass(Class<?> classToProcess) {
+        MethodHandleSpec methodHandleSpec = methodHandleSpecFor(classToProcess);
+        var existing = seedSpecs.put(classToProcess, methodHandleSpec);
+        if (existing != null) {
+            throw new IllegalArgumentException("class " + classToProcess.getSimpleName() + "has already been added");
+        }
+        return this;
+    }
+
+    /**
+     * Equivalent to multiple chained calls to {@link #addClass}.
+     */
+    public Injector addClasses(Class<?>... classesToProcess) {
+        return addClasses(Arrays.asList(classesToProcess));
+    }
+
+    /**
+     * Equivalent to multiple chained calls to {@link #addClass}.
+     */
+    public Injector addClasses(Collection<Class<?>> classesToProcess) {
+        classesToProcess.forEach(this::addClass);
+        return this;
+    }
+
+    /**
+     * Equivalent to {@link #addInstance addInstance(object.getClass(), object)}.
+     */
+    public <T> Injector addInstance(Object object) {
+        @SuppressWarnings("unchecked")
+        Class<T> actualClass = (Class<T>) object.getClass(); // Whatever the runtime type is, it's represented by T
+        return addInstance(actualClass, actualClass.cast(object));
+    }
+
+    /**
+     * Equivalent to multiple calls to {@link #addInstance(Object)}.
+     */
+    public Injector addInstances(Object... objects) {
+        for (var x : objects) {
+            addInstance(x);
+        }
+        return this;
+    }
+
+    /**
+     * Equivalent to multiple calls to {@link #addInstance(Object)}.
+     */
+    public Injector addInstances(Iterable<?> objects) {
+        for (var x : objects) {
+            addInstance(x);
+        }
+        return this;
+    }
+
+    /**
+     * Indicates that <code>object</code> is to be injected for parameters of type <code>type</code>.
+     * The given object is treated as though it had been instantiated by the injector.
+     */
+    public <T> Injector addInstance(Class<? super T> type, T object) {
+        assert type.isInstance(object); // No unchecked casting shenanigans allowed
+        var existing = seedSpecs.put(type, new ExistingInstanceSpec(type, object));
+        if (existing != null) {
+            throw new IllegalStateException("There's already an object for " + type);
+        }
+        return this;
+    }
+
+    /**
+     * For each "component" (getter) <em>c</em> of a {@link Record}, calls {@link #addInstance(Class, Object)} to register the
+     * value with the component's declared type.
+     */
+    @SuppressForbidden(reason = "Can't call invokeExact because we don't know the exact Record subtype statically")
+    public <T> Injector addRecordContents(Record r) {
```

Comment:
This was inspired by `PluginServiceInstances`. I can remove it if you want.

Reply:
I think we should be explicit about what instances we want to add into the injector.

Reply:
`PluginServiceInstances` changes pretty frequently. Every few weeks I get a merge conflict there.

Reply:
I don't think changes to that class should really affect this, they should be independent. I am suggesting the instances we want to pass to the injector are explicitly added with `addInstance` in `NodeConstructor`. Changes to `PluginServiceInstances` shouldn't break that statically, it would only leave us possibly without a new instance in the injector, but that is trivial to fix, and shouldn't cause a merge conflict.

Reply:
Ok I get it. We don't necessarily want `NodeConstruction` to have 19 `addInstance` calls; but the point is that we don't want to offer `addRecordContents` as part of the `Injector` API because then it will start to be adopted and we'll have to support it forever.

I've moved it to `NodeConstruction`.